keps/sig-node/4639-oci-volume-source/README.md

```diff
@@ -0,0 +1,964 @@
+Signoff checklist to be completed.
+
+Check these off as they are completed for the Release Team to track. These
+checklist items _must_ be updated for the enhancement to be released.
+-->
+
+Items marked with (R) are required *prior to targeting to a milestone / release*.
+
+- [ ] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input (including test refactors)
+  - [ ] e2e Tests for all Beta API Operations (endpoints)
+  - [ ] (R) Ensure GA e2e tests meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+The proposed enhancement adds a new `VolumeSource` to Kubernetes that supports OCI images and/or OCI artifacts.
+This allows users to package files and share them among containers in a pod without including them in the main image,
+thereby reducing vulnerabilities and simplifying image creation.
+
+While OCI images are well-supported by Kubernetes and CRI,
+extending support to OCI artifacts involves recognizing additional media types, implementing custom lifecycle management,
+and ensuring appropriate validation and security measures.
+
+## Motivation
+
+Supporting OCI images and artifacts directly as a `VolumeSource` allows
+Kubernetes to focus on OCI standards as well as allows to store and distribute
+any content using OCI registries. This allows the project to grow into use cases
+which go beyond running particular images.
+
+### Goals
+
+- Introduce a new `VolumeSource` type that allows mounting OCI images and/or artifacts.
```

Comment:
Images with layers have a well known media type for the layers, and defined behaviors for expanding the collection of tar layers into an overlay filesystem. But artifacts can have any media type, not necessarily a tar with optional compression. How would unknown media types be handled, and would it be possible to extract a single file directly stored as a single layer with a specified media type?

For example, wasm style artifacts may include the runnable binary directly as a layer, or a ML model may have their own binary format for packaging their model that doesn't require additional wrapping with tar or gzip.

Reply:
Brandon, this VolumeSource type is for the kubernetes API, vs an artifact type.  I agree, may have mentioned, but not to sufficient detail yet, the issue with IANA types possibly having their own definitions for if and how they should be "mounted."

Reply:
> . How would unknown media types be handled, and would it be possible to extract a single file directly stored as a single layer with a specified media type?

Isn't it more that we would need to explicitly define media types that *are* handled?  I am not seeing much in the text about the media type handling beyond
>  Update container runtimes to recognize and handle new media types associated with OCI artifacts.

But I think this needs to be explicitly specified.

I think there's going to be a split between:

"I want to embed this single file at the target path" which would map to the AI model case, as well as other things like "I want to inspect this ISO image which is stored as an OCI artifact".
vs
"I have a filesystem tree" use case....which is actually a lot like what ConfigMaps are.

Reply:
For the media types that are handled, make sure that also covers the scenario with multiple layers. There are some assumptions when you go from one tar file to multiple tar files that you'll want to document.

For the unknown media types, I think the options are to either reject, ignore/skip, or allow the user to map specific media types to specific files. The latter would also need to handle multiple layer scenarios.

This needs to avoid any design that would change behavior as media types change from unknown to known in later releases (you don't want a file to suddenly become a filesystem in a new release). So a user supplied mapping could override any known media type actions.

Reply:
Would there be a benefit to supporting a media type + an artifact type (+ platform)?

Suppose you have an Image Index which references multiple Image Manifests which themselves reference multi-layer blobs. Each of these Image Manifests could potentially have unique platforms/artifactTypes representing different configurations.

Distributing artifacts in this way would be beneficial as it would enable the grouping of conceptually similar artifacts together. This would be potentially beneficial from a storage perspective as well because it would enable caching of each individual layers while still enabling an entire filesystem to be represented.

Reply:
Sometimes the owner/author of AI/ML model is different from the image, and the model may be updated independent from the image. model as a shared layer will lead to challenge for online training and model updating, if layer is updated, the images(which share the model) have to update too.

Reply:
I intentionally left open / removed the media type discussion because:

- there is no standard yet and creating one takes time (tried that for a seccomp profile a while ago).
- runtimes can still decide which media types to accept or reject
- we can improve this KEP and the API later on if a standard exists